tidy3d/plugins/adjoint/components/medium.py

```diff
@@ -0,0 +1,287 @@
+"""Defines jax-compatible mediums."""
+from __future__ import annotations
+
+from typing import Dict, Tuple, Union
+
+import pydantic as pd
+import numpy as np
+import jax.numpy as jnp
+from jax.tree_util import register_pytree_node_class
+
+from ....components.types import Bound, Literal
+from ....components.medium import Medium, AnisotropicMedium, CustomMedium
+from ....components.data.monitor_data import FieldData
+from ....components.data.dataset import PermittivityDataset
+from ....components.data.data_array import ScalarFieldDataArray
+
+from .base import JaxObject
+from .types import JaxFloat, validate_jax_float
+from .data.data_array import JaxDataArray
+from .data.dataset import JaxPermittivityDataset
+
+# number of integration points per unit wavelength in material
+PTS_PER_WVL_INTEGRATION = 20
+
+
+@register_pytree_node_class
+class JaxMedium(Medium, JaxObject):
+    """A :class:`.Medium` registered with jax."""
+
+    permittivity: JaxFloat = pd.Field(
+        1.0,
+        title="Permittivity",
+        description="Relative permittivity of the medium. May be a ``jax`` ``DeviceArray``.",
+        jax_field=True,
+    )
+
+    _sanitize_permittivity = validate_jax_float("permittivity")
+
+    def to_medium(self) -> Medium:
+        """Convert :class:`.JaxMedium` instance to :class:`.Medium`"""
+        self_dict = self.dict(exclude={"type"}).copy()
+        return Medium.parse_obj(self_dict)
+
+    # pylint: disable =too-many-locals
+    def _get_volume_disc(
+        self, grad_data: FieldData, sim_bounds: Bound, wvl_mat: float
+    ) -> Tuple[Dict[str, np.ndarray], float]:
+        """Get the coordinates and volume element for the inside of the corresponding structure."""
+
+        # find intersecting volume between structure and simulation
+        mnt_bounds = grad_data.monitor.geometry.bounds
+        rmin, rmax = self.bounds_intersection(mnt_bounds, sim_bounds)
+
+        # assemble volume coordinates and differential volume element
+        d_vol = 1.0
```

Comment:
Yea, this brings up some good points. A few things:
* the adjoint plugin currently assumes that the geometry being modeled by the structure is sufficiently resolved the FDTD solver to be treated analytically. Practically speaking, it means that we tend to integrate over the surfaces and volumes described by the `Structure.geometry` rather than the actual geometry on the grid. This path was chosen out of simplicity because it becomes pretty annoying to handle annoying edge cases that can pop up here, but could lead to errors in the gradient if the resolution is too low. 
* Currently we have a single constant defining the grids per wavelength in the material used by the integration schemes. But as you mention, maybe if the field variation is very high, we might need to adaptively choose this based in the grid.
* Quick note: this `JaxMedium` vip code actually only works for `JaxStructures` with `JaxBox` geometry. As I add the `JaxPolyslab`, my thinking is to introduce an extra step in this integration where we evaluate the integrand only at points that are `.inside()` the geometry.

I think youre right that we could consider a similar integration scheme using the raw coordinates, I plan to work on this a bit later though, since practically speaking, this scheme is simpler and yields very good results already.

Reply:
Sounds good